src/main/java/bio/terra/common/kubernetes/KubeService.java

```diff
@@ -0,0 +1,190 @@
+
+    logger.info(
+        "Kubernetes configuration: inKube: "
+            + inKubernetes
+            + "; namespace: "
+            + namespace
+            + "; podName: "
+            + podName);
+  }
+
+  /**
+   * Get a list of the API pods from Kubernetes. It is returned as a set to make probing for
+   * specific names easy.
+   *
+   * @return set of pod names containing the podNameFilter string; null if not in kubernetes
+   */
+  public Set<String> getPodList() {
+    Set<String> pods = new HashSet<>();
+    if (!inKubernetes) {
+      return pods;
+    }
+
+    try {
+      CoreV1Api api = makeCoreApi();
+      V1PodList list =
+          api.listNamespacedPod(namespace, null, null, null, null, null, null, null, null, null);
+      for (V1Pod item : list.getItems()) {
+        if (item.getMetadata() != null) {
+          String podName = item.getMetadata().getName();
+          if (StringUtils.contains(podName, podNameFilter)) {
+            pods.add(podName);
+          }
+        }
+      }
+      return pods;
+    } catch (ApiException ex) {
+      throw new KubeApiException("Error listing pods", ex);
+    }
+  }
+
+  /**
+   * Launch the {@link KubePodListener} thread.
+   *
+   * @param stairway the Stairway instance passed to the pod listener for Stairway recovery
+   */
+  public void startPodListener(Stairway stairway) {
+    if (inKubernetes) {
+      podListener = new KubePodListener(this, stairway, namespace, podNameFilter);
+      podListenerThread = new Thread(podListener);
+      podListenerThread.start();
+    }
+  }
+
+  public int getActivePodCount() {
+    if (podListener != null) {
+      return podListener.getActivePodCount();
+    }
+    int defaultPodCount = 1;
+    logger.info("KubeService ActivePodCount - default val: {}", defaultPodCount);
+    return defaultPodCount;
+  }
+
+  /**
+   * Stop pod listener flips the isShutdown flag so we don't do it more than once. It tries to join
+   * the pod listener thread.
+   *
+   * @param timeUnit time unit to wait for the listener thread to stop
+   * @param joinWait time in the time unit to wait
+   * @return true if we joined the listener thread; false if that timed out
+   */
+  public boolean stopPodListener(TimeUnit timeUnit, long joinWait) {
+    if (!isShutdown.get() && inKubernetes) {
+      isShutdown.set(true);
+      // stop the pod listener thread
+      podListenerThread.interrupt();
+      long waitMillis = timeUnit.toMillis(joinWait);
+      try {
+        podListenerThread.join(waitMillis);
+      } catch (InterruptedException ex) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public String getPodName() {
+    return podName;
+  }
+
+  public String getNamespace() {
+    return namespace;
+  }
+
+  // Method for pod listener to test shutdown state
+  boolean isShutdown() {
+    return isShutdown.get();
+  }
+
+  private CoreV1Api makeCoreApi() {
+    try {
+      ApiClient client = ClientBuilder.cluster().build();
```

Comment:
This is confusing me. You make one kind of client and then return a different one?

Reply:
This is how the K8s API manages its different features. You build different sub-clients for each thing you want to do. So there are alpha clients and beta clients. All kinds of clients.